File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-ird-5-4/FallbackReferencePrice.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class FallbackReferencePrice : ISerialized
    {
        public FallbackReferencePrice(XmlNode xmlNode)
        {
            XmlNode valuationPostponementNode = xmlNode.SelectSingleNode("valuationPostponement");
            
            if (valuationPostponementNode != null)
            {
                if (valuationPostponementNode.Attributes["href"] != null || valuationPostponementNode.Attributes["id"] != null) 
                {
                    if (valuationPostponementNode.Attributes["id"] != null) 
                    {
                        valuationPostponementIDRef_ = valuationPostponementNode.Attributes["id"].Value;
                        ValuationPostponement ob = new ValuationPostponement(valuationPostponementNode);
                        IDManager.SetID(valuationPostponementIDRef_, ob);
                    }
                    else if (valuationPostponementNode.Attributes["href"] != null)
                    {
                        valuationPostponementIDRef_ = valuationPostponementNode.Attributes["href"].Value;
                    }
                    else
                    {
                        valuationPostponement_ = new ValuationPostponement(valuationPostponementNode);
                    }
                }
                else
                {
                    valuationPostponement_ = new ValuationPostponement(valuationPostponementNode);
                }
            }
            
        
            XmlNodeList fallbackSettlementRateOptionNodeList = xmlNode.SelectNodes("fallbackSettlementRateOption");
            
            if (fallbackSettlementRateOptionNodeList != null)
            {
                this.fallbackSettlementRateOption_ = new List<SettlementRateOption>();
                foreach (XmlNode item in fallbackSettlementRateOptionNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            fallbackSettlementRateOptionIDRef_ = item.Attributes["id"].Value;
                            fallbackSettlementRateOption_.Add(new SettlementRateOption(item));
                            IDManager.SetID(fallbackSettlementRateOptionIDRef_, fallbackSettlementRateOption_[fallbackSettlementRateOption_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            fallbackSettlementRateOptionIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        fallbackSettlementRateOption_.Add(new SettlementRateOption(item));
                        }
                    }
                    else
                    {
                        fallbackSettlementRateOption_.Add(new SettlementRateOption(item));
                    }
                }
            }
            
        
            XmlNode fallbackSurveyValuationPostponenmentNode = xmlNode.SelectSingleNode("fallbackSurveyValuationPostponenment");
            
            if (fallbackSurveyValuationPostponenmentNode != null)
            {
                if (fallbackSurveyValuationPostponenmentNode.Attributes["href"] != null || fallbackSurveyValuationPostponenmentNode.Attributes["id"] != null) 
                {
                    if (fallbackSurveyValuationPostponenmentNode.Attributes["id"] != null) 
                    {
                        fallbackSurveyValuationPostponenmentIDRef_ = fallbackSurveyValuationPostponenmentNode.Attributes["id"].Value;
                        Empty ob = new Empty(fallbackSurveyValuationPostponenmentNode);
                        IDManager.SetID(fallbackSurveyValuationPostponenmentIDRef_, ob);
                    }
                    else if (fallbackSurveyValuationPostponenmentNode.Attributes["href"] != null)
                    {
                        fallbackSurveyValuationPostponenmentIDRef_ = fallbackSurveyValuationPostponenmentNode.Attributes["href"].Value;
                    }
                    else
                    {
                        fallbackSurveyValuationPostponenment_ = new Empty(fallbackSurveyValuationPostponenmentNode);
                    }
                }
                else
                {
                    fallbackSurveyValuationPostponenment_ = new Empty(fallbackSurveyValuationPostponenmentNode);
                }
            }
            
        
            XmlNode calculationAgentDeterminationNode = xmlNode.SelectSingleNode("calculationAgentDetermination");
            
            if (calculationAgentDeterminationNode != null)
            {
                if (calculationAgentDeterminationNode.Attributes["href"] != null || calculationAgentDeterminationNode.Attributes["id"] != null) 
                {
                    if (calculationAgentDeterminationNode.Attributes["id"] != null) 
                    {
                        calculationAgentDeterminationIDRef_ = calculationAgentDeterminationNode.Attributes["id"].Value;
                        CalculationAgent ob = new CalculationAgent(calculationAgentDeterminationNode);
                        IDManager.SetID(calculationAgentDeterminationIDRef_, ob);
                    }
                    else if (calculationAgentDeterminationNode.Attributes["href"] != null)
                    {
                        calculationAgentDeterminationIDRef_ = calculationAgentDeterminationNode.Attributes["href"].Value;
                    }
                    else
                    {
                        calculationAgentDetermination_ = new CalculationAgent(calculationAgentDeterminationNode);
                    }
                }
                else
                {
                    calculationAgentDetermination_ = new CalculationAgent(calculationAgentDeterminationNode);
                }
            }
            
        
        }
        
    
        #region ValuationPostponement_
        private ValuationPostponement valuationPostponement_;
        public ValuationPostponement ValuationPostponement_
        {
            get
            {
                if (this.valuationPostponement_ != null)
                {
                    return this.valuationPostponement_; 
                }
                else if (this.valuationPostponementIDRef_ != null)
                {
                    valuationPostponement_ = IDManager.getID(valuationPostponementIDRef_) as ValuationPostponement;
                    return this.valuationPostponement_; 
                }
                else
                {
                      return this.valuationPostponement_; 
                }
            }
            set
            {
                if (this.valuationPostponement_ != value)
                {
                    this.valuationPostponement_ = value;
                }
            }
        }
        #endregion
        
        public string valuationPostponementIDRef_ { get; set; }
        #region FallbackSettlementRateOption_
        private List<SettlementRateOption> fallbackSettlementRateOption_;
        public List<SettlementRateOption> FallbackSettlementRateOption_
        {
            get
            {
                if (this.fallbackSettlementRateOption_ != null)
                {
                    return this.fallbackSettlementRateOption_; 
                }
                else if (this.fallbackSettlementRateOptionIDRef_ != null)
                {
                    return this.fallbackSettlementRateOption_; 
                }
                else
                {
                      return this.fallbackSettlementRateOption_; 
                }
            }
            set
            {
                if (this.fallbackSettlementRateOption_ != value)
                {
                    this.fallbackSettlementRateOption_ = value;
                }
            }
        }
        #endregion
        
        public string fallbackSettlementRateOptionIDRef_ { get; set; }
        #region FallbackSurveyValuationPostponenment_
        private Empty fallbackSurveyValuationPostponenment_;
        public Empty FallbackSurveyValuationPostponenment_
        {
            get
            {
                if (this.fallbackSurveyValuationPostponenment_ != null)
                {
                    return this.fallbackSurveyValuationPostponenment_; 
                }
                else if (this.fallbackSurveyValuationPostponenmentIDRef_ != null)
                {
                    fallbackSurveyValuationPostponenment_ = IDManager.getID(fallbackSurveyValuationPostponenmentIDRef_) as Empty;
                    return this.fallbackSurveyValuationPostponenment_; 
                }
                else
                {
                      return this.fallbackSurveyValuationPostponenment_; 
                }
            }
            set
            {
                if (this.fallbackSurveyValuationPostponenment_ != value)
                {
                    this.fallbackSurveyValuationPostponenment_ = value;
                }
            }
        }
        #endregion
        
        public string fallbackSurveyValuationPostponenmentIDRef_ { get; set; }
        #region CalculationAgentDetermination_
        private CalculationAgent calculationAgentDetermination_;
        public CalculationAgent CalculationAgentDetermination_
        {
            get
            {
                if (this.calculationAgentDetermination_ != null)
                {
                    return this.calculationAgentDetermination_; 
                }
                else if (this.calculationAgentDeterminationIDRef_ != null)
                {
                    calculationAgentDetermination_ = IDManager.getID(calculationAgentDeterminationIDRef_) as CalculationAgent;
                    return this.calculationAgentDetermination_; 
                }
                else
                {
                      return this.calculationAgentDetermination_; 
                }
            }
            set
            {
                if (this.calculationAgentDetermination_ != value)
                {
                    this.calculationAgentDetermination_ = value;
                }
            }
        }
        #endregion
        
        public string calculationAgentDeterminationIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-ird-5-4/FallbackReferencePrice.cpp
// FallbackReferencePrice.cpp 
#include "FallbackReferencePrice.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

FallbackReferencePrice::FallbackReferencePrice(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //valuationPostponementNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* valuationPostponementNode = xmlNode->FirstChildElement("valuationPostponement");

   if(valuationPostponementNode){valuationPostponementIsNull_ = false;}
   else{valuationPostponementIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- valuationPostponementNode , address : " << valuationPostponementNode << std::endl;
   #endif
   if(valuationPostponementNode)
   {
      if(valuationPostponementNode->Attribute("href") || valuationPostponementNode->Attribute("id"))
      {
          if(valuationPostponementNode->Attribute("id"))
          {
             valuationPostponementIDRef_ = valuationPostponementNode->Attribute("id");
             valuationPostponement_ = boost::shared_ptr<ValuationPostponement>(new ValuationPostponement(valuationPostponementNode));
             valuationPostponement_->setID(valuationPostponementIDRef_);
             IDManager::instance().setID(valuationPostponementIDRef_,valuationPostponement_);
          }
          else if(valuationPostponementNode->Attribute("href")) { valuationPostponementIDRef_ = valuationPostponementNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { valuationPostponement_ = boost::shared_ptr<ValuationPostponement>(new ValuationPostponement(valuationPostponementNode));}
   }

   //fallbackSettlementRateOptionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fallbackSettlementRateOptionNode = xmlNode->FirstChildElement("fallbackSettlementRateOption");

   if(fallbackSettlementRateOptionNode){fallbackSettlementRateOptionIsNull_ = false;}
   else{fallbackSettlementRateOptionIsNull_ = true;}

   if(fallbackSettlementRateOptionNode)
   {
      for(fallbackSettlementRateOptionNode; fallbackSettlementRateOptionNode; fallbackSettlementRateOptionNode = fallbackSettlementRateOptionNode->NextSiblingElement("fallbackSettlementRateOption")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fallbackSettlementRateOptionNode , address : " << fallbackSettlementRateOptionNode << std::endl;
          #endif
          if(fallbackSettlementRateOptionNode->Attribute("href") || fallbackSettlementRateOptionNode->Attribute("id"))
          {
              if(fallbackSettlementRateOptionNode->Attribute("id"))
              {
                  fallbackSettlementRateOptionIDRef_ = fallbackSettlementRateOptionNode->Attribute("id");
                  fallbackSettlementRateOption_.push_back(boost::shared_ptr<SettlementRateOption>(new SettlementRateOption(fallbackSettlementRateOptionNode)));
                  fallbackSettlementRateOption_.back()->setID(fallbackSettlementRateOptionIDRef_);
                  IDManager::instance().setID(fallbackSettlementRateOptionIDRef_, fallbackSettlementRateOption_.back());
              }
              else if(fallbackSettlementRateOptionNode->Attribute("href")) { fallbackSettlementRateOptionIDRef_ = fallbackSettlementRateOptionNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { fallbackSettlementRateOption_.push_back(boost::shared_ptr<SettlementRateOption>(new SettlementRateOption(fallbackSettlementRateOptionNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fallbackSettlementRateOptionNode , address : " << fallbackSettlementRateOptionNode << std::endl;
       #endif
   }

   //fallbackSurveyValuationPostponenmentNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fallbackSurveyValuationPostponenmentNode = xmlNode->FirstChildElement("fallbackSurveyValuationPostponenment");

   if(fallbackSurveyValuationPostponenmentNode){fallbackSurveyValuationPostponenmentIsNull_ = false;}
   else{fallbackSurveyValuationPostponenmentIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fallbackSurveyValuationPostponenmentNode , address : " << fallbackSurveyValuationPostponenmentNode << std::endl;
   #endif
   if(fallbackSurveyValuationPostponenmentNode)
   {
      if(fallbackSurveyValuationPostponenmentNode->Attribute("href") || fallbackSurveyValuationPostponenmentNode->Attribute("id"))
      {
          if(fallbackSurveyValuationPostponenmentNode->Attribute("id"))
          {
             fallbackSurveyValuationPostponenmentIDRef_ = fallbackSurveyValuationPostponenmentNode->Attribute("id");
             fallbackSurveyValuationPostponenment_ = boost::shared_ptr<Empty>(new Empty(fallbackSurveyValuationPostponenmentNode));
             fallbackSurveyValuationPostponenment_->setID(fallbackSurveyValuationPostponenmentIDRef_);
             IDManager::instance().setID(fallbackSurveyValuationPostponenmentIDRef_,fallbackSurveyValuationPostponenment_);
          }
          else if(fallbackSurveyValuationPostponenmentNode->Attribute("href")) { fallbackSurveyValuationPostponenmentIDRef_ = fallbackSurveyValuationPostponenmentNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { fallbackSurveyValuationPostponenment_ = boost::shared_ptr<Empty>(new Empty(fallbackSurveyValuationPostponenmentNode));}
   }

   //calculationAgentDeterminationNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* calculationAgentDeterminationNode = xmlNode->FirstChildElement("calculationAgentDetermination");

   if(calculationAgentDeterminationNode){calculationAgentDeterminationIsNull_ = false;}
   else{calculationAgentDeterminationIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- calculationAgentDeterminationNode , address : " << calculationAgentDeterminationNode << std::endl;
   #endif
   if(calculationAgentDeterminationNode)
   {
      if(calculationAgentDeterminationNode->Attribute("href") || calculationAgentDeterminationNode->Attribute("id"))
      {
          if(calculationAgentDeterminationNode->Attribute("id"))
          {
             calculationAgentDeterminationIDRef_ = calculationAgentDeterminationNode->Attribute("id");
             calculationAgentDetermination_ = boost::shared_ptr<CalculationAgent>(new CalculationAgent(calculationAgentDeterminationNode));
             calculationAgentDetermination_->setID(calculationAgentDeterminationIDRef_);
             IDManager::instance().setID(calculationAgentDeterminationIDRef_,calculationAgentDetermination_);
          }
          else if(calculationAgentDeterminationNode->Attribute("href")) { calculationAgentDeterminationIDRef_ = calculationAgentDeterminationNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { calculationAgentDetermination_ = boost::shared_ptr<CalculationAgent>(new CalculationAgent(calculationAgentDeterminationNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<ValuationPostponement> FallbackReferencePrice::getValuationPostponement()
{
   if(!this->valuationPostponementIsNull_){
        if(valuationPostponementIDRef_ != ""){
             return boost::shared_static_cast<ValuationPostponement>(IDManager::instance().getID(valuationPostponementIDRef_));
        }else{
             return this->valuationPostponement_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ValuationPostponement>();
   }
}
std::vector<boost::shared_ptr<SettlementRateOption>> FallbackReferencePrice::getFallbackSettlementRateOption()
{
   if(!this->fallbackSettlementRateOptionIsNull_){
        if(fallbackSettlementRateOptionIDRef_ != ""){
             return this->fallbackSettlementRateOption_;
        }else{
             return this->fallbackSettlementRateOption_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<SettlementRateOption>>();
   }
}
boost::shared_ptr<Empty> FallbackReferencePrice::getFallbackSurveyValuationPostponenment()
{
   if(!this->fallbackSurveyValuationPostponenmentIsNull_){
        if(fallbackSurveyValuationPostponenmentIDRef_ != ""){
             return boost::shared_static_cast<Empty>(IDManager::instance().getID(fallbackSurveyValuationPostponenmentIDRef_));
        }else{
             return this->fallbackSurveyValuationPostponenment_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Empty>();
   }
}
boost::shared_ptr<CalculationAgent> FallbackReferencePrice::getCalculationAgentDetermination()
{
   if(!this->calculationAgentDeterminationIsNull_){
        if(calculationAgentDeterminationIDRef_ != ""){
             return boost::shared_static_cast<CalculationAgent>(IDManager::instance().getID(calculationAgentDeterminationIDRef_));
        }else{
             return this->calculationAgentDetermination_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CalculationAgent>();
   }
}
}

